File: lib/src/main/java/net/imoya/android/fragment/BaseFragment.java
package net.imoya.android.fragment;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;

import net.imoya.android.util.Log;

/**
 * ナビゲーション前提の {@link Fragment} です。
 * {@link BaseActivity} と組み合わせて使用することを想定しています。
 */
public abstract class BaseFragment extends Fragment {
    protected static final String ARGUMENT_CONTAINER_VIEW_ID =
            BaseFragment.class.getName() + ".containerViewId";

    /**
     * {@link Application} の {@link Context}
     */
    protected Context appContext = null;

    private static final String TAG = "BaseFragment";

    /**
     * {@link Activity} に於いて、この {@link Fragment} が配置される親 {@link View} の
     * IDを設定します。
     * {@link BaseFragment} の実行前にこのメソッドを呼び出すことにより、
     * {@link #replaceThisTo(BaseFragment, String)} メソッドを利用可能となります。
     *
     * @param fragment        ID を設定する {@link BaseFragment}
     * @param containerViewId 親 {@link View} のID
     */
    public static void setContainerViewId(@NonNull BaseFragment fragment, int containerViewId) {
        final Bundle argumentsBefore = fragment.getArguments();
        final Bundle arguments = (argumentsBefore != null ? argumentsBefore : new Bundle());
        arguments.putInt(ARGUMENT_CONTAINER_VIEW_ID, containerViewId);
        if (argumentsBefore == null) {
            fragment.setArguments(arguments);
        }
    }

    /**
     * {@link Activity} に於いて、この {@link Fragment} が配置される親 {@link View} の
     * IDを取得します。
     *
     * @return 親 {@link View} のID
     */
    protected static int getContainerViewId(@NonNull BaseFragment fragment) {
        return fragment.requireArguments().getInt(ARGUMENT_CONTAINER_VIEW_ID, 0);
    }

    /**
     * この {@link Fragment} が配置されている場所へ、別の {@link Fragment} を配置します。
     * <p/>
     * 画面の切り替え時、次のアニメーションを使用します。
     * アニメーションの内容をカスタマイズする場合は、
     * アプリケーション側のリソースで、下記のリソースを再定義してください:
     * <ul>
     * <li>{@link R.anim#fragment_enter}</li>
     * <li>{@link R.anim#fragment_exit}</li>
     * <li>{@link R.anim#fragment_pop_enter}</li>
     * <li>{@link R.anim#fragment_pop_exit}</li>
     * </ul>
     *
     * @param fragment 置き換え後の {@link Fragment}
     * @param tag      {@link Fragment} に設定するタグ
     */
    protected void replaceThisTo(@NonNull BaseFragment fragment, @Nullable String tag) {
        final int containerViewId = BaseFragment.getContainerViewId(this);
        setContainerViewId(fragment, containerViewId);
        this.getParentFragmentManager().beginTransaction()
                .setCustomAnimations(
                        R.anim.fragment_enter, R.anim.fragment_exit,
                        R.anim.fragment_pop_enter, R.anim.fragment_pop_exit)
                .replace(containerViewId, fragment, tag)
                .addToBackStack(tag)
                .commit();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.appContext = context.getApplicationContext();
    }

    @SuppressWarnings("deprecation")
    @Override
    public void onAttach(@NonNull Activity activity) {
        super.onAttach(activity);
        this.appContext = activity.getApplicationContext();
    }

    /**
     * {@link ActionBar} の表示非表示を設定します。
     *
     * @param visible 表示する場合true, 非表示とする場合false
     */
    @SuppressWarnings("SameParameterValue")
    protected void setActionBarVisibility(boolean visible) {
        final Activity activity = this.getActivity();
        Log.d(TAG, "setActionBarVisibility: activity = " + activity);
        if (activity instanceof AppCompatActivity) {
            final ActionBar actionBar = ((AppCompatActivity) activity).getSupportActionBar();
            Log.d(TAG, "setActionBarVisibility: actionBar = " + actionBar);
            if (actionBar != null) {
                if (visible) {
                    actionBar.show();
                } else {
                    actionBar.hide();
                }
            }
        }
    }

    /**
     * アクションバー上のタイトル文字列を設定します。
     *
     * @param title タイトル文字列
     */
    protected void setTitle(String title) {
        final Activity activity = this.getActivity();
        if (activity != null) {
            activity.setTitle(title);
        }
    }

    /**
     * アクションバー上のタイトル文字列を設定します。
     *
     * @param titleId タイトル文字列のリソースID
     */
    protected void setTitle(@StringRes int titleId) {
        this.setTitle(this.requireContext().getString(titleId));
    }
}

File: lib/src/main/java/net/imoya/android/fragment/BaseFragment.kt
/*
 * Copyright (C) 2022 IceImo-P
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.imoya.android.fragment

import android.app.ActionBar
import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment

/**
 * ナビゲーション前提の [Fragment] です。
 * [BaseActivity] と組み合わせて使用することを想定しています。
 */
@Suppress("unused")
abstract class BaseFragment : Fragment() {
    /**
     * [Application] の [Context]
     */
    @Suppress("MemberVisibilityCanBePrivate")
    protected lateinit var appContext: Context

    /**
     * この [Fragment] が配置されている場所へ、別の [Fragment] を配置します。
     *
     * 画面の切り替え時、次のアニメーションを使用します。
     * アニメーションの内容をカスタマイズする場合は、
     * アプリケーション側のリソースで下記のリソースを再定義してください:
     *
     *  * [R.anim.fragment_enter]
     *  * [R.anim.fragment_exit]
     *  * [R.anim.fragment_pop_enter]
     *  * [R.anim.fragment_pop_exit]
     *
     * @param fragment 置き換え後の [Fragment]
     * @param tag      [Fragment] に設定するタグ
     */
    protected open fun replaceThisTo(fragment: BaseFragment, tag: String?) {
        val containerViewId = this.containerViewId
        fragment.containerViewId = containerViewId
        parentFragmentManager.beginTransaction()
            .setCustomAnimations(
                R.anim.fragment_enter, R.anim.fragment_exit,
                R.anim.fragment_pop_enter, R.anim.fragment_pop_exit
            )
            .replace(containerViewId, fragment, tag)
            .addToBackStack(tag)
            .commit()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        appContext = context.applicationContext
    }

    @Suppress("deprecation")
    @Deprecated("Deprecated in Java")
    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        appContext = activity.applicationContext
    }

    /**
     * [ActionBar] の表示非表示を設定します。
     *
     * @param visible 表示する場合true, 非表示とする場合false
     */
    protected open fun setActionBarVisibility(visible: Boolean) {
        val activity: Activity? = this.activity
        FragmentLog.v(TAG) { "setActionBarVisibility: activity = $activity" }
        if (activity is AppCompatActivity) {
            val actionBar = activity.supportActionBar
            FragmentLog.v(TAG) { "setActionBarVisibility: actionBar = $actionBar" }
            if (actionBar != null) {
                if (visible) {
                    actionBar.show()
                } else {
                    actionBar.hide()
                }
            }
        }
    }

    /**
     * アクションバー上のタイトル文字列を設定します。
     *
     * @param title タイトル文字列
     */
    @Suppress("MemberVisibilityCanBePrivate")
    protected fun setTitle(title: CharSequence?) {
        this.activity?.title = title ?: ""
    }

    /**
     * アクションバー上のタイトル文字列を設定します。
     *
     * @param titleId タイトル文字列のリソースID
     */
    protected fun setTitle(@StringRes titleId: Int) {
        this.setTitle(requireContext().getString(titleId))
    }

    /**
     * [Activity] に於いて、この [Fragment] が配置される親 [View] の ID
     */
    var containerViewId: Int
        get() = requireArguments().getInt(ARGUMENT_CONTAINER_VIEW_ID, 0)
        set(containerViewId) {
            val argumentsBefore = this.arguments
            val arguments = argumentsBefore ?: Bundle()
            arguments.putInt(ARGUMENT_CONTAINER_VIEW_ID, containerViewId)
            if (argumentsBefore == null) {
                setArguments(arguments)
            }
        }

    companion object {
        /**
         * Arguments key: この [Fragment] が配置される親 [View] の ID
         */
        protected val ARGUMENT_CONTAINER_VIEW_ID =
            BaseFragment::class.java.name + ".containerViewId"

        /**
         * Tag for log
         */
        private const val TAG = "BaseFragment"
    }
}